src/Interpreters/Cache/FileCache.cpp

```diff
@@ -497,62 +497,68 @@ FileCache::FileSegmentCell * FileCache::addCell(
     {
-        FileSegment::State result_state = state;
-        if (state == FileSegment::State::EMPTY && enable_cache_hits_threshold)
+        auto record = stash_records.find({key, offset});
+
+        if (record == stash_records.end())
         {
-            auto record = stash_records.find({key, offset});
+            auto priority_iter = stash_priority->add(key, offset, 0, cache_lock);
+            stash_records.insert({{key, offset}, priority_iter});
 
-            if (record == stash_records.end())
+            if (stash_priority->getElementsNum(cache_lock) > max_stash_element_size)
             {
-                auto priority_iter = stash_priority->add(key, offset, 0, cache_lock);
-                stash_records.insert({{key, offset}, priority_iter});
-
-                if (stash_priority->getElementsNum(cache_lock) > max_stash_element_size)
-                {
-                    auto remove_priority_iter = stash_priority->getLowestPriorityWriteIterator(cache_lock);
-                    stash_records.erase({remove_priority_iter->key(), remove_priority_iter->offset()});
-                    remove_priority_iter->removeAndGetNext(cache_lock);
-                }
-
-                /// For segments that do not reach the download threshold,
-                /// we do not download them, but directly read them
-                result_state = FileSegment::State::SKIP_CACHE;
+                auto remove_priority_iter = stash_priority->getLowestPriorityWriteIterator(cache_lock);
+                stash_records.erase({remove_priority_iter->key(), remove_priority_iter->offset()});
+                remove_priority_iter->removeAndGetNext(cache_lock);
             }
-            else
-            {
-                auto priority_iter = record->second;
-                priority_iter->use(cache_lock);
 
-                result_state = priority_iter->hits() >= enable_cache_hits_threshold
-                    ? FileSegment::State::EMPTY
-                    : FileSegment::State::SKIP_CACHE;
-            }
+            /// For segments that do not reach the download threshold,
+            /// we do not download them, but directly read them
+            result_state = FileSegment::State::SKIP_CACHE;
         }
+        else
+        {
+            auto priority_iter = record->second;
+            priority_iter->use(cache_lock);
 
-        return std::make_shared<FileSegment>(offset, size, key, this, result_state, settings);
-    };
+            result_state = priority_iter->hits() >= enable_cache_hits_threshold
+                ? FileSegment::State::EMPTY
+                : FileSegment::State::SKIP_CACHE;
+        }
+    }
 
-    FileSegmentCell cell(skip_or_download(), this, cache_lock);
     auto & offsets = files[key];
-
     if (offsets.empty())
     {
         auto key_path = getPathInLocalCache(key);
 
         if (!fs::exists(key_path))
-            fs::create_directories(key_path);
+        {
+            try
```

Comment:
It makes sense to add a comment about the possible reasons - why sometimes creating a directory fails and why we should catch the exception here instead of letting it propagate and be caught in the upper layer.

Reply:
ok, will add a comment, the initial reason for this PR was that create_directories can fail with "no space left on device". And it is less convenient to propagate exception to the upper level, because we want to skip cache in this case, which should be done here by assigning a file segment the state SKIP_CACHE, which was done in this PR.